### ISSUE TITLE
Moves in guard expressions aren't handled properly

### ISSUE THREAD

**Issue by Aatch** [rust-lang/rust#22073]
This code causes a segfault (without optimisations on):

```rust
use std::fmt;
fn main() {
    let a = std::rc::Rc::new(1);
    
    let x = (1, 2);
    
    match x {
        (_, y) |
        (y, _) if take(y, a) => (),
        _ => ()
    }

}

fn take<T:fmt::Debug>(x: usize, t: T) -> bool {
    println!("{:?}", t);
    x == 1
}
```

as does this similar code:

```rust
use std::fmt;
fn main() {
    let a = std::rc::Rc::new(1);
    
    let x = (1, 2);
    
    match x {
        (_, y) if take(y, a) => (),
        (y, _) if take(y, a) => (),
        _ => ()
    }

}

fn take<T:fmt::Debug>(x: usize, t: T) -> bool {
    println!("{:?}", t);
    x == 1
}
```

This is because the current control flow graph essentially visits each arm in parallel. With multiple patterns, each pattern within the arm is visited, then the guard is visited after all the patterns have been visited, even though this doesn't reflect the real semantics (we visit the pattern, then the guard, then the next pattern, then the guard again).

**Comment by Aatch** [comment]
I'm working on a fix for both of these issues. Should be done fairly soon I hope.

**Comment by pnkfelix** [comment]
@Aatch whoa, I thought we didn't *support* moves into guards in the first place?

(That is, not that I object to you fixing this ... but I had thought there was an explicit decision not to support such guards, at least not in the short term, because we were not confident we could make it work!)

Am I mistaken?

**Comment by Aatch** [comment]
Well I managed to get it to work (you're actually assigned to the PR fixing it).

**Comment by pnkfelix** [comment]
@Aatch ah that is what I was thinking of, indeed.  (My mind isn't made up about whether it makes sense to support moves into guards; if we got away with disallowing mutation, why not moves as well, at least for the short term.)

**Comment by pnkfelix** [comment]
1.0 beta, P-backcompat-lang.

**Comment by pnkfelix** [comment]
closing in favor of #22580 (which is essentially the same as the first commit in this PR, apart from some factoring into separate commits, cutting of some details I decided were not strictly necessary, and slight fixes to the very helpful comment from @Aatch ).
